### ISSUE TITLE
Stronghold.load() is slow

### ISSUE THREAD

**Issue by Iraijus** [tauri-apps/plugins-workspace#2048]
In my JS code, I run the following code:

```
appDataDir()
.then(async dir => {
  console.log(dir)
  const vaultPath = `${dir}/vault.hold`;
  const vaultPassword = 'vault_password';
  const stronghold = await Stronghold.load(vaultPath, vaultPassword);
  console.log("TEST");
})
```

The time between the first and second log is always somewhere between 1 and 2 minutes. Is there a reason why Stronghold.load() is so slow for me?

**Comment by Iraijus** [comment]
Nevermind. I'm still not sure what the issue was, but clearing the app's storage, though I had already tried that before, solved the issue for now.

**Comment by Iraijus** [comment]
Apparently, it only becomes slow after saving a value.

**Comment by UrsDeSwardt** [comment]
I'm having trouble with the underlying `iota_stronghold` being slow as well. Also typically between 1-2 minutes.

**Comment by UrsDeSwardt** [comment]
In case anyone else is struggling with this, I was able to get the underlying `iota_stronghold` working smoothly by downgrading the version to `1.0.5`.

**Comment by lindongchen** [comment]
I also have this issue.

**Comment by gabriel-le** [comment]
It seems that it is only slow for me in debug target, but not release. Is this also the case for you?

**Comment by almeidaz** [comment]
@gabriel-le Yes, it's the same here.

**Comment by Ripwords** [comment]
I went into IOTA's discord and found someone raising up the same issue, one of the members asked to run the program in --release mode, which does makes it faster. But didn't provide any other details

**Comment by ahkohd** [comment]
> I went into IOTA's discord and found someone raising up the same issue, one of the members asked to run the program in --release mode, which does makes it faster. But didn't provide any other details

yes in release mode it's faster.

**Comment by Ripwords** [comment]
Running in release mode does make things a bit more inconvenient.

Just as a workaround, I created a simple class with getKey, setKey and removeKey methods. Using env variables to check if I'm in `DEV` or `PROD` if in DEV I use localStorage just to quickly test without having to wait, and only in PROD will be using stronghold.

**Comment by ahkohd** [comment]
> Running in release mode does make things a bit more inconvenient.
> 
> Just as a workaround, I created a simple class with getKey, setKey and removeKey methods. Using env variables to check if I'm in `DEV` or `PROD` if in DEV I use localStorage just to quickly test without having to wait, and only in PROD will be using stronghold.

yes, this is very neat.

**Comment by kthehatter** [comment]
i'm facing the same issue it works fine in release mode but too slow in development mode using @Ripwords method seems to be a temporary fix to help focus on other things

**Comment by xiaohulu** [comment]
I am facing the same issue too, too slow in development mode

**Comment by endrik-exe** [comment]
I also noticed something strange, the memory keeps increasing by factor of 20-30 MB, about 16 times (480mb) : everytime i called the load method.

![Image](https://github.com/user-attachments/assets/59d7fee7-4e73-4901-b1df-d28c44cf923f)

**Comment by HuakunShen** [comment]
> I also noticed something strange, the memory keeps increasing by factor of 20-30 MB, about 16 times (480mb) : everytime i called the load method.
> 
> ![Image](https://github.com/user-attachments/assets/59d7fee7-4e73-4901-b1df-d28c44cf923f)

@endrik-exe is this in dev mode or prod mode? If prod I will need to remove it from my project.

**Comment by endrik-exe** [comment]
@HuakunShen This is dev mode, haven't tested it in prod yet.

**Comment by Deano217** [comment]
Does anyone know what the underlying problem is? Since this was tagged as upstream, I checked the [issues](https://github.com/iotaledger/stronghold.rs/issues) on the stronghold.rs repo, but couldn't find anything related.

**Comment by GeorgeGkas** [comment]
I've confirmed that this issue occurs only in debug mode, everything works fine in release. While it's just speculation, the root cause may be related to unoptimized cryptographic operations (encryption/decryption). Given that both the `load()` and `save()` methods are affected, this seems like a reasonable assumption. The thing is that this impacts negatively the development experience.

**Comment by yunger7** [comment]
Issue seems to be with the `scrypt` crate, which is a dependency of `iota-crypto`. Try adding this to `Cargo.toml`

```toml
[profile.dev.package.scrypt]
opt-level = 3
```

**Comment by yunger7** [comment]
Issue seems to be with the `scrypt` crate, which is a dependency of `iota-crypto`. Try adding this to `Cargo.toml`

```toml
[profile.dev.package.scrypt]
opt-level = 3
```

**Comment by Max-Loo** [comment]
> Issue seems to be with the `scrypt` crate, which is a dependency of `iota-crypto`. Try adding this to `Cargo.toml`
> 
> [profile.dev.package.scrypt]
> opt-level = 3

Works for my too!

**Comment by Guanyuhao** [comment]
> Issue seems to be with the `scrypt` crate, which is a dependency of `iota-crypto`. Try adding this to `Cargo.toml`
> 
> [profile.dev.package.scrypt]
> opt-level = 3

Works for my too! +1

**Comment by FabianLars** [comment]
Added the profile to the readme and plugin guide on the website. I doubt we'll see upstream fixes and we cannot define the opt-level in the plugin either (cargo ignores profile configs in dependencies). Also a headsup: the stronghold plugin will be deprecated soon (it obviously already is but we didn't make it official yet)

**Comment by sionzee** [comment]
> Added the profile to the readme and plugin guide on the website. I doubt we'll see upstream fixes and we cannot define the opt-level in the plugin either (cargo ignores profile configs in dependencies). Also a headsup: the stronghold plugin will be deprecated soon (it obviously already is but we didn't make it official yet)

I was about to start implementing Stronghold, but I see it's being deprecated. What's the recommended replacement?

**Comment by FabianLars** [comment]
For most use cases we believe using the keychain (or keyring) is a better choice. I was working on an official plugin (the PR should still be there) but there are so many plugins for this already and using the keyring-rs crates directly is also so easy that I'm not sure if there's really much value in yet another keyring plugin 🤔

And if you want to keep using stronghold you can still use the rust crates instead of our plugin.

**Comment by mike-lloyd03** [comment]
> the stronghold plugin will be deprecated soon (it obviously already is but we didn't make it official yet)

What makes it obvious that it is already deprecated? I wouldn't have known this was the case if I hadn't found this comment.

Just spent the better part of the last two hours in my app retooling everything to make use of the stronghold plugin only to find this.

@FabianLars if the plan is to no longer support this plugin, can you please put a note on the tauri website?

**Comment by FabianLars** [comment]
Yes you're absolutely right, and it's totally my bad. I didn't anticipate that the replacement plugin would take so long and I didn't adapt to the timeline changes.
I'll see to update the website soon and not wait for the replacement, which was the initial plan.

**Comment by mike-lloyd03** [comment]
Thank you! For what it's worth, I'm working with this right now:

https://github.com/charlesportwoodii/tauri-plugin-keyring
